### ISSUE TITLE
Strengthen constraints in DB

### ISSUE THREAD

**Issue by aefhm** [oasisprotocol/oasis-indexer#50]
We can add several constraints to the DB to catch errors early:
- foreign key constraints, e.g. for `beneficiary` or `delegatee`
- narrower type checks, e.g. `tx_hash TEXT CHECK (length(tx_hash)=64)` or `round BIGINT CHECK (round>=0)`

**Comment by mitjat** [comment]
This will be easier once we make FK constrains deferrable (see e.g. https://www.alibabacloud.com/blog/postgresql-deferrable-constraints-unique-primary-key-foreign-key-and-exclude_597717)

**Comment by mitjat** [comment]
It would also be useful (for data consistency and readability) to introduce [custom types](https://www.postgresql.org/docs/current/sql-createdomain.html). Example:
```
CREATE DOMAIN oasis_addr TEXT CHECK(length(VALUE) = 46 AND VALUE ~ '^oasis1')
```

**Comment by mitjat** [comment]
Note: I closed #195 but when we introduce deferred constraints we should eyeball error logs and make sure that that violated-FK log entry is gone.

Also, raising priority to p2 because of problems we're having with inconsistent state; stronger checks might help us find the cause sooner.

**Comment by mitjat** [comment]
https://app.clickup.com/t/3ua4u66

**Comment by mitjat** [comment]
Resolved by #217
